### PR TITLE
ICU-20469 Adding userguide for span fields and data build tool.

### PR DESCRIPTION
**Pull request by sffc**
Also creates new directory to store userguide markdown files.

##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20138
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [x] Tests included
- [x] Documentation is changed or added

### CODE REVIEW

**Comment by sffc** [comment]
Incomplete because I am waiting on #413 before being able to finish the Java section of the new userguide page.

**Comment by sffc** [comment]
I've written up userguide pages in Markdown for FormattedValue and for the data build tool.  Ready for review.

File: docs/userguide/data_buildtool.md [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+
+### File Slicing (coarse-grained features)
+
+ICU provides a lot of features, of which you probably need only a small subset for your application.  Feature slicing is a powerful way to prune out data for any features you are not using.
+
+***CAUTION:*** When slicing by features, you must manually include all dependencies.  For example, if you are formatting dates, you must include not only the date formatting data but also the number formatting data, since dates contain numbers.  Expect to spend a fair bit of time debugging your feature filter to get it to work the way you expect it to.
+
+The data for many ICU features live in individual files.  ICU Data Build Tool puts similar *types* of files into categories.  The following table summarizes the ICU data files and their corresponding features and categories:
+
+Feature | Category ID(s) | Data Files <br/> ([icu4c/source/data](https://github.com/unicode-org/icu/tree/master/icu4c/source/data))
```
Comment (jefgen) [inline docs/userguide/data_buildtool.md]: More of a minor nit: the markdown table seems to render a bit mangled on GitHub.
(Was it exported from something perhaps?)

Comment (sffc) [reply docs/userguide/data_buildtool.md]: I hand-wrote the table. It renders correctly in my local Markdown renderer (non-GFM).  I'll look into it.

Comment (sffc) [reply docs/userguide/data_buildtool.md]: I fixed the table for GitHub and added a column showing file sizes.

File: docs/userguide/data_buildtool.md [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+        rules: [
+          -/units
+          -/unitsNarrow
+          -/unitsShort
+          +/*/length
+        ]
+      }
+    ]
+
+At this time, the wildcard character must stand alone in a piece of the path.
```
Comment (markusicu) [inline docs/userguide/data_buildtool.md]: Reading this, I was confused whether "At this time" referred to the flow of this document.

How about "The wildcard must be the only character in its path segment. Future ICU versions may expand the syntax."

File: docs/userguide/data_buildtool.md [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+
+In accordance with *filterType* "locale", the parent locales *en* and *root* would get both units; this is required since both en-US and en-CA may inherit from the parent locale:
+
+    -/units
+    -/unitsNarrow
+    -/unitsShort
+    +/*/length/mile
+    +/*/length/kilometer
+
+## Other Features of ICU Data Build Tool
```
Comment (markusicu) [inline docs/userguide/data_buildtool.md]: "the"

File: docs/userguide/data_buildtool.md [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+
+While data filtering is the primary reason the ICU Data Build Tool was developed, there are other features you can use it for, too.
```
Comment (markusicu) [inline docs/userguide/data_buildtool.md]: ... there are additional use cases.

File: docs/userguide/data_buildtool.md [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+
+### Running Data Build without Configure/Make
+
+You can build the dat file outside of the ICU build system by directly invoking the Python buildtool.  Run the following command to see the help text for the CLI tool:
+
+    $ PYTHONPATH=path/to/icu4c/source/data python3 -m buildtool --help
+
+### Collation UCAData
```
Comment (markusicu) [inline docs/userguide/data_buildtool.md]: Misplaced section -- this is still part of the filtering.

File: docs/userguide/data_buildtool.md [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+
+There are two versions of the Collation data file:
+
+- ucadata-unihan.txt
+- ucadata-implicithan.txt
+
+By default, the unihan version is used.  To use the implicithan version, pass the following command line argument to buildtool:
```
Comment (markusicu) [inline docs/userguide/data_buildtool.md]: We need to say what the differences are and why someone might want to choose one vs. the other.

How about

The unihan version sorts Han characters in radical-stroke order according to Unicode, which is a somewhat useful default sort order, especially for use with non-CJK languages. The implicithan version sorts Han characters in the order of their Unicode assignment, which is similar to radical-stroke order for common characters but arbitrary for others. The unihan version of the data file is much larger than that for implicithan, so if you need collation but also small data, then you may want to select the implicithan version.

Include sample data file sizes for the two versions in ICU 64.
Include link: https://www.unicode.org/reports/tr10/#Implicit_Weights

File: docs/userguide/data_buildtool.md [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+
+### Running Data Build without Configure/Make
+
+You can build the dat file outside of the ICU build system by directly invoking the Python buildtool.  Run the following command to see the help text for the CLI tool:
+
+    $ PYTHONPATH=path/to/icu4c/source/data python3 -m buildtool --help
+
+### Collation UCAData
+
+There are two versions of the Collation data file:
```
Comment (markusicu) [inline docs/userguide/data_buildtool.md]: For using collation (sorting & searching) in any language, the "root" collation data file must be included. It provides the Unicode CLDR default sort order for all code points, and forms the basis for language-specific tailorings as well as for custom collators built at runtime.

There are two versions of the root collation data file:

File: docs/userguide/formatted_value.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+```cpp
+// Let fmt be a DateIntervalFormat for locale en-US and skeleton dMMMMy
+// Let input1 be July 20, 2018 and input2 be August 3, 2018:
+FormattedDateInterval result = fmt->formatToValue(*input1, *input2, status);
+assertEquals("Expected output from format",
+    u"July 20 \u2013 August 3, 2018", result.toString(status));
+ConstrainedFieldPosition cfpos;
+cfpos.constrainField(UFIELD_CATEGORY_DATE_INTERVAL_SPAN, 0);
+if (result.nextPosition(cfpos, status)) {
+    assertEquals("Expect start index", cfpos.getStart(), 0);
```
Comment (markusicu) [inline docs/userguide/formatted_value.md]: Nit: Usually, the expected value is first.

File: docs/userguide/data_buildtool.md [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+
+The ICU Data Build Tool will work out of the box with a default Python installation.  In order to use Hjson syntax, the `hjson` pip module must be installed on your system.  You should also consider installing the `jsonschema` module to print messages when errors are found in your config file.
+
+    $ pip3 install --user hjson jsonschema
+
+To build ICU4J with custom data, you must first build ICU4C with custom data and then generate the JAR file.  For more information, read [icu4j-readme.txt](https://github.com/unicode-org/icu/blob/master/icu4c/source/data/icu4j-readme.txt).
+
+### Locale Slicing
+
+The simplest way to slice ICU data is by locale.  ICU Data Build Tool makes it easy to select your desired locales to suit a number of use cases.
```
[the remark this answers is not in the document]
Comment (sffc) [reply docs/userguide/data_buildtool.md]: Done

File: docs/userguide/data_buildtool.md [hunk 10]
```diff
@@ -0,0 +1,380 @@
+ICU Data Build Tool
+===================
+
+ICU 64 provides a tool for configuring your ICU locale data file with greater granularity.  This page explains how to use this tool to customize and reduce your data file size.
+
+## Overview: What is in the ICU data file?
+
+ICU uses a data bundle file named something like `icudt64l.dat`, which is little-endian data for ICU 64. This dat file is packaged into the `libicudata.so` on Linux or `libicudata.dll.a` on Windows.  In ICU4J, it is bundled into a jar file named `icudata.jar`.
+
+There are hundreds of **locales** supported in ICU (including script and region variants).  Each of those locales contains data for one or more ICU **features**.  At a high level, the size of the ICU data file corresponds to the cross-product of locales and features, except that not all locales contain data for all features, and there are some features that are general and don't have locale-specific data.
```
[the remark this answers is not in the document]
Comment (sffc) [reply docs/userguide/data_buildtool.md]: I took this comment and rephrased this section.

File: docs/userguide/data_buildtool.md [hunk 11]
```diff
@@ -0,0 +1,380 @@
...
+      }
+    }
+
+Do *not* include directories or file extensions.  They will be added automatically for you.  Note that all files in a particular category have the same directory and extension.
+
+You can use either a whitelist or a blacklist for the file name filter.
+
+##### Regex Filter
+
+To exclude filenames matching a certain regular expression, use *filterType* "regex".  For example, to reject the CJK-specific break iteration rules:
```
[the remark this answers is not in the document]
Comment (sffc) [reply docs/userguide/data_buildtool.md]: Done

File: docs/userguide/data_buildtool.md [hunk 12]
```diff
@@ -0,0 +1,380 @@
...
+    featureFilters:
+      curr_tree: {
+        filterType: locale
+        whitelist: [
+          it
+        ]
+      }
+    }
+
+It may be common to simply exclude an entire `_tree` category without affecting other categories.  For example, to exclude region display names:
```
[the remark this answers is not in the document]
Comment (sffc) [reply docs/userguide/data_buildtool.md]: Done

File: docs/userguide/data_buildtool.md (hunk 2 above)
Comment (sffc) [reply docs/userguide/data_buildtool.md]: Thanks; done

File: docs/userguide/data_buildtool.md (hunk 3 above)
Comment (sffc) [reply docs/userguide/data_buildtool.md]: Done

File: docs/userguide/data_buildtool.md (hunk 4 above)
Comment (sffc) [reply docs/userguide/data_buildtool.md]: Done

File: docs/userguide/data_buildtool.md (hunk 6 above)
Comment (sffc) [reply docs/userguide/data_buildtool.md]: Thank you for this!

File: docs/userguide/data_buildtool.md (hunk 7 above)
Comment (sffc) [reply docs/userguide/data_buildtool.md]: Thank you for this!

File: docs/userguide/data_buildtool.md (hunk 5 above)
Comment (sffc) [reply docs/userguide/data_buildtool.md]: Could go either place.  For now, this is a separate option for buildtool, so it makes sense to have it down here.  When https://unicode-org.atlassian.net/browse/ICU-20473 is fixed, the section should be moved up.  I made a note of it on that ticket.

File: docs/userguide/formatted_value.md (hunk 8 above)
Comment (sffc) [reply docs/userguide/formatted_value.md]: Good catch!

File: docs/userguide/data_buildtool.md [hunk 13]
```diff
@@ -1,39 +1,67 @@
 ICU Data Build Tool
```
Comment (markusicu) [inline docs/userguide/data_buildtool.md]: Question about file names and organization:

The User Guide so far has used multiple folders (pages with sub-pages) to organize its contents.
https://sites.google.com/site/icuprojectuserguide/system/app/pages/sitemap/hierarchy

Should we recreate this here, or should we use a flat set of files, maybe with topic prefixes like "data_" here?

In Sites, we only have an "icudata" page, but we could decide to make a "data/" folder with pages like overview.md and buildtool.md.

Would a folder structure be a problem for other tools, like automatic index pages or navigation panes?

Comment (sffc) [reply docs/userguide/data_buildtool.md]: Good thoughts.  My opinion is, let's defer this decision to when the rest of the userguide migration takes place.

**Pull request merged**